enhancements/openstack-baremetal-compute.md

```diff
@@ -0,0 +1,141 @@
+---
+title: openstack-baremetal-compute
+authors:
+  - "@pierreprinetti"
+reviewers:
+  - "@openshift/team-openstack-cloud-provider"
+approvers:
+  - TBD
+creation-date: 2019-12-10
+last-updated: 2019-12-10
+status: provisional
+
+---
+
+# Openstack Baremetal Compute
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+OpenStack's [Nova][openstack-nova] service can be configured to provision
+baremetal machines as OpenShift Compute Nodes.
+
+This enhancement will allow [Ironic][openstack-ironic]-provisioned machines to
+be used as OpenShift Compute Nodes, when they are made available as Nova
+flavors.
+
+## Motivation
+
+In heavy-loaded clusters, or when the workload is better handled with specially
+purposed hardware, the functional requirement of controlling where the work is
+computed might outweight the added value of virtualisation.
+
+In OpenStack clusters where Nova transparently provisions baremetal machines,
+this enhancement will ensure that OpenShift installations can use them as
+Compute Nodes.
+
+### Goals
+
+* Enable running OpenShift on a mix of virtual and baremetal machines.
+* Baremetal machines can be added and removed from a running OpenShift cluster.
+
+### Non-Goals
+
+An OpenStack setting where baremetal machines can't be attached to a virtual
+subnet is not supported in this enhancement.
+
+## Proposal
+
+This enhancement is less about adding a new logical piece, than it is about
+uncovering unknowns in a configuration that might work out-of-the-box in
+selected environments.
+
+Therefore, the implementation consists of:
+* identifying a reference architecture
+* anticipating probable issues
+* testing and fixing errors
+
+### Characteristics of the reference architecture
+
+* The reference OpenStack cluster transparently provisions baremetal machines
+* Provisioned baremetal machines can be attached to a Neutron virtual subnet
```

Comment:
I'd be a little more explicit here

* The choice between deploying a VM or BM node is done by selecting an appropriate Nova flavour
  - which can be done both in install config and passed to Cluster API provider OpenStack
  - we would (for the time being) *not* expose this in the interactive prompt -- you'd edit install-config.yaml
* Both deployment requests make the same API calls to Nova (no explicit calls to Ironic or anything in the Installer required)
  - so creating BM nodes should work transparently from the installer as well as CAPO
* both VMs and BM consume the same image from glance (I'm not 100% sure that's actually, the case but at least the BM folks will have something to latch on to :-).
* uncertainty about the more extra networking resources we use: ports (including trunk), security groups, allowed address pairs?
   - are all these necessary for BM (they might not be)?
   - will things break if the installer or CAPO tries to set them on BM nodes? (they might because it's an unsupported operation or something?)
    - e.g. can you even assign a port or a SG to a BM node?
* We should explicitly mention any specific OpenStack services and configuration this "transparent BM provisioning" will require. There's going to be Ironic and Nova config at least and I'm sure a fair bit of Neutron too.
  - we probably won't do a full tripleo config or whatnot
  - but at least hit the keywords so people can actually read it and understand what needs to happen

Reply:
> We should explicitly mention any specific OpenStack services and configuration

That's why I've come up with a "reference architecture" section, but I ended up adding unrelated content in that section. I'll have to rework it. Thanks for bringing that point up